### PR TITLE
Charge density in Source/Drain/Channel

### PR DESCRIPTION
**Pull request by prkkumar**
This PR adds the option to model the source, drain, and channel regions within the overall SC region by modifying the formula for charge density.

We now use the general form of rho as 
`rho = q(p - n - N_a  + N_d)` where 
`p` is hole density
`n` is electron density
`N_a` and `N_d` are the concentration of ionized acceptor and donor, respectively.

In the current implementation, channel is assumed to have non zero acceptor doping (p-type) and source/drain have non-zero donor doping (n-type) [based on  [Girish's paper](https://ieeexplore.ieee.org/document/9833284)

### CODE REVIEW

**Comment by jackieyao0114** [comment]
Not so relevant to this PR, but I think the following statement got reversed:
`n the current implementation, channel is assumed to have non zero acceptor doping (n-type) and source/drain have non-zero donor doping (p-type)`
p-type should be acceptor, n type is a donor.

**Comment by jackieyao0114** [comment]
Is it possible to visualize the charge densities just like how you did it in the paper and quantify the doping effect?

**Comment by prkkumar** [comment]
> Not so relevant to this PR, but I think the following statement got reversed: `n the current implementation, channel is assumed to have non zero acceptor doping (n-type) and source/drain have non-zero donor doping (p-type)` p-type should be acceptor, n type is a donor.

Thanks, yes that was a mistake. Fixed it in the first comment.

File: Source/FerroX.cpp [hunk 1]
```diff
@@ -321,6 +325,18 @@ void InitializeFerroXNamespace() {
          }
      }
 
+     if (pp.queryarr("Channel_lo",temp)) {
```
Comment (ajnonaka) [inline Source/FerroX.cpp:328]: General comment: do you want any of these queryarr calls to be getarr instead?  Queryarr is optional and typically you supply default values for query/querryarr (not needed for getarr)

File: Source/Solver/ChargeDensity.cpp [hunk 2]
```diff
@@ -14,21 +14,29 @@ void ComputeRho(MultiFab&      PoissonPhi,
...
+
 
         amrex::ParallelFor( bx, [=] AMREX_GPU_DEVICE (int i, int j, int k)
         {
+	     Real x = prob_lo[0] + (i+0.5) * dx[0];
+             Real y = prob_lo[1] + (j+0.5) * dx[1];
              Real z = prob_lo[2] + (k+0.5) * dx[2];
 
-             if(z <= SC_hi[2]){ //SC region
+             if (x <= SC_hi[0] + small && x >= SC_lo[0] - small && y <= SC_hi[1] + small && y >= SC_lo[1] - small && z <= SC_hi[2] + small && z >= SC_lo[2] - small) {
```
Comment (ajnonaka) [inline Source/Solver/ChargeDensity.cpp:39]: I imagine the general usage of the code you define regions so entire cells are of that material, in which case the "small" corrections are only absolutely needed for face or nodal data.  Up to you if you keep it this way or remove them, just commenting :)

Comment (ajnonaka) [reply Source/Solver/ChargeDensity.cpp:39]: I suppose this is irrelevant once the mask PR comes into play.

**Review by ajnonaka — APPROVED** [review]
Approve, pending the comment about query vs get

File: Source/Solver/ChargeDensity.cpp (hunk 2 above)
Comment (prkkumar) [reply Source/Solver/ChargeDensity.cpp:39]: Yes, this goes away in the next PR :)

File: Source/FerroX.cpp (hunk 1 above)
Comment (prkkumar) [reply Source/FerroX.cpp:328]: Thanks! This has been addressed in PR#26. 

I made FE_lo/hi to be getarr. Everything else is still queryarr, but I have added default values for everything that's obtained by queryarr. The inline comment explains the logic behind the default values and why everything but FE_lo/hi is an optional parameter. We can make changes if this logic doesn't make sense or if we have other preferences.

**Pull request merged**